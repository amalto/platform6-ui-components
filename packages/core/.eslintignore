dist/
loader/

src/components.d.ts