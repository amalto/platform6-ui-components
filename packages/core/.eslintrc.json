{
  "extends": ["../../.eslintrc.json", "plugin:@stencil/recommended", "plugin:jest/recommended", "plugin:jsx-a11y/recommended"],
  "parserOptions": {
    "project": ["./tsconfig.json", "./tsconfig.lint.json"],
    "ecmaFeatures": {
      "jsx": true // Allows for the parsing of JSX
    }
  },
  "plugins": ["jest"],

  "rules": {
    "jsx-a11y/label-has-associated-control": [
      2,
      {
        "assert": "either"
      }
    ],
    "react/jsx-props-no-spreading": "off",
    "react/no-unknown-property": [2, { "ignore": ["class"] }]
  },
  "settings": {
    "react": {
      "pragma": "h"
    }
  }
}
