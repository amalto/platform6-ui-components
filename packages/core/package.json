{
  "name": "@platform6/components",
  "version": "0.0.1",
  "description": "Platform 6 UI components",
  "repository": {
    "type": "git",
    "url": "https://github.com/amalto/platform6-ui-components"
  },
  "license": "MIT",
  "main": "dist/index.cjs.js",
  "module": "dist/custom-elements/index.js",
  "collection": "dist/collection/collection-manifest.json",
  "collection:main": "dist/collection/index.js",
  "es2015": "dist/esm/index.mjs",
  "es2017": "dist/esm/index.mjs",
  "unpkg": "dist/stencil-component/stencil-component.js",
  "types": "dist/types/index.d.ts",
  "files": [
    "dist/",
    "loader/"
  ],
  "scripts": {
    "build": "stencil build --docs",
    "format": "prettier --write ./",
    "postformat": "npm run lint.fix",
    "generate": "stencil generate",
    "lint": "npm run lint.ts && npm run lint.sass",
    "lint.fix": "npm run lint.ts.fix && npm run lint.sass.fix",
    "lint.sass": "stylelint 'src/**/*.scss'",
    "lint.sass.fix": "npm run lint.sass -- --fix",
    "lint.ts": "eslint --cache --ext .js,.jsx,.ts,.tsx ./",
    "lint.ts.fix": "npm run lint.ts -- --fix",
    "start": "stencil build --docs --dev --watch --serve",
    "test": "stencil test --spec --e2e",
    "test.watch": "stencil test --spec --e2e --watchAll",
    "storybook": "start-storybook -p 6006 -s ./dist --ci",
    "build-storybook": "build-storybook"
  },
  "dependencies": {
    "@creativebulma/bulma-tooltip": "^1.2.0",
    "@fortawesome/fontawesome-svg-core": "^1.2.32",
    "@fortawesome/free-brands-svg-icons": "^5.15.1",
    "@fortawesome/free-regular-svg-icons": "^5.15.1",
    "@fortawesome/free-solid-svg-icons": "^5.15.1",
    "@stencil/core": "2.2.0",
    "bulma": "^0.9.1",
    "bulma-calendar": "^6.0.9",
    "choices.js": "9.0.1",
    "lit-html": "^1.3.0"
  },
  "devDependencies": {
    "@babel/core": "^7.12.3",
    "@stencil/eslint-plugin": "^0.3.1",
    "@stencil/postcss": "^1.0.1",
    "@stencil/react-output-target": "0.0.9",
    "@stencil/sass": "^1.3.2",
    "@storybook/addon-a11y": "^6.0.28",
    "@storybook/addon-actions": "^6.0.28",
    "@storybook/addon-essentials": "^6.0.28",
    "@storybook/addon-links": "^6.0.28",
    "@storybook/html": "^6.0.28",
    "@storybook/preset-scss": "^1.0.3",
    "@storybook/theming": "^6.0.28",
    "@storybook/web-components": "^6.0.28",
    "@types/jest": "^26.0.15",
    "@types/js-beautify": "^1.11.0",
    "@typescript-eslint/eslint-plugin": "^4.6.1",
    "@typescript-eslint/parser": "^4.6.1",
    "autoprefixer": "^9.8.6",
    "babel-loader": "^8.1.0",
    "css-loader": "^5.0.0",
    "eslint-plugin-jest": "^24.1.0",
    "jest": "^26.6.3",
    "jest-cli": "^26.6.3",
    "js-beautify": "^1.13.0",
    "prettier": "^2.1.2",
    "puppeteer": "^5.4.1",
    "react-is": "^17.0.1",
    "sass-loader": "^10.0.5",
    "style-loader": "^2.0.0",
    "typescript": "4.0.5"
  }
}
