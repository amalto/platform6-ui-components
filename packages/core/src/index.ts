export { Components, JSX } from './components';
export { P6ButtonType } from './components/atoms/p6-button/p6-button';
export { P6CalendarType } from './components/atoms/p6-calendar/p6-calendar';
export { P6CheckboxValue } from './components/atoms/p6-checkbox/p6-checkbox';
export { P6InputValue } from './components/atoms/p6-input/p6-input';
export { P6InputType } from './components/atoms/p6-input/types';
export { Target } from './components/atoms/p6-link/p6-link';
export { P6RadioValue } from './components/atoms/p6-radio/p6-radio';
export { P6SelectControl, P6SelectValue } from './components/atoms/p6-select-native/p6-select-native';
export { P6SwitchValue } from './components/atoms/p6-switch/p6-switch';
export { P6TextareaValue } from './components/atoms/p6-textarea/p6-textarea';
export { P6LanguageValue } from './components/molecules/p6-language/p6-language';
export { P6TranslationControl, P6TranslationValue } from './components/molecules/p6-translation/p6-translation';
export {
  AlignColumnDetail,
  CellValueChangedDetail,
  Column,
  ColumnDefinition,
  ColumnId,
  DataItem,
  EditingCellDetail,
  FilterRowsDetail,
  MoveColumnDetail,
  ResetDefinitionsDetail,
  ResizeColumnDetail,
  Row,
  RowId,
  ShowColumnDetail,
  ShowOptionsDetail,
  SortColumnDetail,
} from './components/organisms/p6-grid/core/entities';
export { ServiceStatus } from './components/templates/p6-service-panel/service-status.enum';
export { P6Control } from './shared/form/control';
export { InvalidEventDetail, isCustomEvent, isInvalidEvent, isValidEvent, ValidEventDetail } from './shared/form/event';
export { Alignment, Mode, modes, Position, positions, Size, sizes, SortOrder } from './shared/types';
export { EnumArrayEntry, enumArrayToObject, enumToArray, EnumType } from './utils/enum';
export { LanguageCode } from './utils/language';
