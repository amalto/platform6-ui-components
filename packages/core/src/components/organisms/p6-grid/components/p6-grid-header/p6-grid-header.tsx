import { Component, Element, h, Host, JSX } from '@stencil/core';

@Component({
  tag: 'p6-grid-header',
  styleUrl: 'p6-grid-header.scss',
  shadow: true,
})
export class P6GridHeader {
  @Element() host!: HTMLP6GridHeaderElement;

  render(): JSX.Element {
    return (
      <Host>
        <slot />
      </Host>
    );
  }
}
