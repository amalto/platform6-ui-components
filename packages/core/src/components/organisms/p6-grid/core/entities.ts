import { EventEmitter, JSX } from '@stencil/core';
import { Alignment, Direction, SortOrder } from '../../../../shared/types';

export type ColumnId = string;

export type DataItem = Record<string, unknown>;

export interface ColumnDefinition<T extends DataItem> {
  id?: ColumnId;
  label: string;
  field?: string | keyof T;
  editable?: boolean;
  filtreable?: boolean;
  sortable?: boolean;
  hidden?: boolean;
  width?: number;
  align?: Alignment;
  color?: string;
  sortOrder?: SortOrder;
  disableHeaderMenu?: boolean;
  getValue?: (data: T, context: ColumnDefinition<T>) => string;
  setValue?: (value: string, data: T, context: ColumnDefinition<T>) => T;
  cellEditor?: (data: T, context: ColumnDefinition<T>) => JSX.Element & CellEditor<T>;
  cellRenderer?: (data: T, rowId: RowId, context: ColumnDefinition<T>) => string | HTMLElement;
}

export interface Column<T extends DataItem> extends ColumnDefinition<T> {
  id: ColumnId;
  editable: boolean;
  hidden: boolean;
  width: number;
  align: Alignment;
  sortOrder: SortOrder;
  disableHeaderMenu: boolean;
  getValue: (data: T, context: ColumnDefinition<T>) => string;
  setValue: (value: string, data: T, context: ColumnDefinition<T>) => T;
}

export interface CellEditor<T extends DataItem> {
  column: Column<T>;
  row: Row<T>;
  value: string;

  p6GridCellValueChanged: EventEmitter<CellValueChangedDetail<T>>;
  p6GridEditingCell: EventEmitter<EditingCellDetail<T>>;
}
export type RowId = string;

export interface Row<T extends DataItem> {
  id: RowId;
  data: T;
}

interface ColumnEventDetail<T extends DataItem> {
  column: Column<T>;
}

export interface MoveColumnDetail<T extends DataItem> extends ColumnEventDetail<T> {
  direction: Direction;
}
export type ResizeColumnDetail<T extends DataItem> = ColumnEventDetail<T>;

export type AlignColumnDetail<T extends DataItem> = ColumnEventDetail<T>;

export type SortColumnDetail<T extends DataItem> = ColumnEventDetail<T>;

export type ShowColumnDetail<T extends DataItem> = ColumnEventDetail<T>;

export interface CellValueChangedDetail<T extends DataItem> {
  column: Column<T>;
  row: Row<T>;
  oldValue: string;
}

export enum EditingCellStatus {
  Start,
  Stop,
  Cancel,
}

export interface EditingCellDetail<T extends DataItem> {
  column: Column<T>;
  row: Row<T>;
  status: EditingCellStatus;
}

export interface ResetDefinitionsDetail {
  reset: boolean;
}

export interface ShowOptionsDetail {
  visible: boolean;
}

export interface FilterRowsDetail {
  value: string;
}
