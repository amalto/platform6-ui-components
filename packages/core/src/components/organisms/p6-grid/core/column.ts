import { Alignment, Direction, SortOrder } from '../../../../shared/types';
import { throwError } from '../../../../utils/error';
import { Column, ColumnDefinition, ColumnId, DataItem } from './entities';

export const MIN_WIDTH = 75;
export const DEFAULT_WIDTH = MIN_WIDTH;
export const INC_WIDTH = 10;

export function defaultGetValue<T extends DataItem>(data: T, context: ColumnDefinition<T>): string {
  if (context.field === undefined || data[context.field] === undefined) {
    return throwError(`field is undefined or doesn't exist, you should provide a 'getValue' function for ${JSON.stringify(context)}`);
  }

  return `${data[context.field as keyof T]}`;
}

export function defaultSetValue<T extends DataItem>(value: string, data: T, context: ColumnDefinition<T>): T {
  if (context.field === undefined || data[context.field] === undefined) {
    return throwError(`field is undefined or doesn't exist, you should provide a 'setValue' function for ${JSON.stringify(context)}`);
  }

  return {
    ...data,
    [context.field as keyof T]: value,
  };
}

let columnIds = 0;
export function fromDefinition<T extends DataItem>(column: ColumnDefinition<T>): Column<T> {
  // eslint-disable-next-line no-plusplus
  const id = column.id || column.field || columnIds++;

  return {
    ...column,
    id: `col-${id}`,
    editable: column.editable || false,
    hidden: column.hidden || false,
    width: column.width || 100,
    align: column.align || Alignment.center,
    sortOrder: column.sortOrder || SortOrder.none,
    disableHeaderMenu: column.disableHeaderMenu || false,
    getValue: column.getValue || defaultGetValue,
    setValue: column.setValue || defaultSetValue,
  };
}

export function isHidden<T extends DataItem>(column: Column<T>): boolean {
  return !!column.hidden;
}

export function replaceColumn<T extends DataItem>(updatedColumn: Column<T>, allColumns: Column<T>[]): Column<T>[] {
  return allColumns.map(col => (col.id === updatedColumn.id ? updatedColumn : col));
}

export function move<T extends DataItem>(id: ColumnId, direction: Direction, columns: Column<T>[]): Column<T>[] {
  const from = columns.findIndex(col => col.id === id);
  const to = from + (direction === Direction.left ? -1 : 1);

  if (from < 0 || to < 0 || to >= columns.length) {
    return columns;
  }

  const result = columns.slice();
  const item = result.splice(from, 1);

  return result.slice(0, to).concat(item).concat(result.slice(to, columns.length));
}
