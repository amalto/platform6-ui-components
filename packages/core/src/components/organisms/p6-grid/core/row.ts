import { SortOrder } from '../../../../shared/types';
import { Column, DataItem, Row, RowId } from './entities';

let rowIds = 0;
export function fromData<T extends DataItem>(data: T): Row<T> {
  rowIds += 1;

  return {
    id: `row-${rowIds}`,
    data,
  };
}

export function compareRow<T extends DataItem>(column: Column<T> | undefined): (a: Row<T>, b: Row<T>) => number {
  const sortOrder = column?.sortOrder || SortOrder.none;
  const factor = sortOrder === SortOrder.desc ? -1 : 1;

  return (a: Row<T>, b: Row<T>): number => {
    if (column === undefined || column.sortable === false || sortOrder === SortOrder.none) {
      return 0;
    }

    const valueA = column.getValue(a.data, column);
    const valueB = column.getValue(b.data, column);

    return valueA.localeCompare(valueB) * factor;
  };
}

export function filterBy<T extends DataItem>(searchValue: string, columns: Column<T>[]): (row: Row<T>) => boolean {
  return row => columns.filter(col => col.filtreable !== false).some(col => col.getValue(row.data, col).includes(searchValue));
}

export function rangeSelectRow<T extends DataItem>(currentRowIdSelect: RowId, selectedRows: Set<RowId>, displayedRow: Row<T>[]): Set<RowId> {
  const currentSelectedIndex = displayedRow.findIndex(row => row.id === currentRowIdSelect);

  if (currentSelectedIndex === -1) {
    return selectedRows;
  }

  const previousRowIdSelected = Array.from(selectedRows)
    .reverse()
    .find(rowId => displayedRow.findIndex(row => row.id === rowId) > -1);

  if (previousRowIdSelected === undefined) {
    return new Set(selectedRows.add(currentRowIdSelect));
  }

  const previousSelectedIndex = displayedRow.findIndex(row => row.id === previousRowIdSelected);

  let [from, to] = [previousSelectedIndex, currentSelectedIndex];
  let excludedFromSlice = currentRowIdSelect;
  if (currentSelectedIndex < previousSelectedIndex) {
    [from, to] = [to, from];
    excludedFromSlice = previousRowIdSelected;
  }

  return displayedRow
    .slice(from, to)
    .reduce((selected, cur) => {
      return cur.id === undefined ? selected : selected.add(cur.id);
    }, new Set(selectedRows))
    .add(excludedFromSlice);
}

export function replaceRow<T extends DataItem>(updatedRow: Row<T>, allRows: Row<T>[]): Row<T>[] {
  return allRows.map(row => (row.id === updatedRow.id ? updatedRow : row));
}
