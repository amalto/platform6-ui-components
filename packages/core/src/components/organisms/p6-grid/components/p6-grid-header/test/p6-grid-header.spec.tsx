import { newSpecPage } from '@stencil/core/testing';
import { P6GridHeader } from '../p6-grid-header';

describe('p6-grid-header', () => {
  it('renders', async () => {
    const page = await newSpecPage({
      components: [P6GridHeader],
      html: `<p6-grid-header></p6-grid-header>`,
    });
    expect(page.root).toMatchSnapshot();
  });
});
