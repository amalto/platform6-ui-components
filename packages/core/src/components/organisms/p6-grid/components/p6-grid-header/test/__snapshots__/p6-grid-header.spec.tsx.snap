// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`p6-grid-header renders 1`] = `
<p6-grid-header>
  <mock:shadow-root>
    <slot></slot>
  </mock:shadow-root>
</p6-grid-header>
`;
